crates/rust-analyzer/src/main_loop.rs

```diff
@@ -389,19 +399,35 @@ fn loop_turn(
                 req,
             )?,
             Message::Notification(not) => {
-                on_notification(
-                    &connection.sender,
-                    world_state,
-                    &mut loop_state.pending_requests,
-                    &mut loop_state.subscriptions,
-                    not,
-                )?;
+                on_notification(&connection.sender, world_state, loop_state, not)?;
             }
             Message::Response(resp) => {
                 let removed = loop_state.pending_responses.remove(&resp.id);
                 if !removed {
                     log::error!("unexpected response: {:?}", resp)
                 }
+
+                if Some(resp.id) == loop_state.configuration_request_id {
+                    loop_state.configuration_request_id = None;
+                    if let Some(err) = resp.error {
+                        log::error!("failed fetch the server settings: {:?}", err)
+                    } else if resp.result.is_none() {
+                        log::error!("received empty server settings response from the client")
+                    } else {
+                        let new_config =
+                            serde_json::from_value::<Vec<ServerConfig>>(resp.result.unwrap())?
+                                .first()
+                                .expect(
+                                    "The client is expected to always send a non-empty config data",
+                                )
+                                .to_owned();
+                        world_state.update_configuration(
+                            new_config.lru_capacity,
+                            get_options(&new_config, text_document_caps),
+                            get_feature_flags(&new_config, connection),
+                        );
+                    }
```

Comment:
Avoid unwrap?
```suggestion
                    } else if let Some(result) = resp.result {
                        let new_config =
                            serde_json::from_value::<Vec<ServerConfig>>(result)?
                                .first()
                                .expect(
                                    "The client is expected to always send a non-empty config data",
                                )
                                .to_owned();
                        world_state.update_configuration(
                            new_config.lru_capacity,
                            get_options(&new_config, text_document_caps),
                            get_feature_flags(&new_config, connection),
                        );
                    } else {
                        log::error!("received empty server settings response from the client")
                    }
```

Reply:
I check this before unwrapping:

```rust
else if resp.result.is_none() {
    log::error!("received empty server settings response from the client")
}
```

Reply:
Sure, it is clearly seen that you check for `is_none()`, but I prefer not having the `unwrap()` and relying on the type system, so that you avoid the space for the errors like removing that `is_none()` check by-mistake and not checking the code after it. I remember you were talking about avoiding unwraps too https://github.com/rust-analyzer/rust-analyzer/pull/3623#discussion_r393600230
```rust
// Prefer:
if let Some(val) = opt {
    ...
} else {
    ...
}
// To
// 
if opt.is_none() {
    ...
} else {
    let opt = opt.unwrap();
}
```